storage/innobase/btr/btr0btr.cc

```diff
@@ -1269,51 +1270,62 @@ btr_read_autoinc(dict_index_t* index)
 
 /** Read the last used AUTO_INCREMENT value from PAGE_ROOT_AUTO_INC,
 or fall back to MAX(auto_increment_column).
-@param[in]	table	table containing an AUTO_INCREMENT column
-@param[in]	col_no	index of the AUTO_INCREMENT column
-@return	the AUTO_INCREMENT value
-@retval	0 on error or if no AUTO_INCREMENT value was used yet */
-ib_uint64_t
-btr_read_autoinc_with_fallback(const dict_table_t* table, unsigned col_no)
+@param table          table containing an AUTO_INCREMENT column
+@param col_no         index of the AUTO_INCREMENT column
+@param mysql_version  TABLE_SHARE::mysql_version
+@param max            the maximum value of the AUTO_INCREMENT column
+@return the AUTO_INCREMENT value
+@retval 0 on error or if no AUTO_INCREMENT value was used yet */
+uint64_t btr_read_autoinc_with_fallback(const dict_table_t *table,
+                                        unsigned col_no, ulong mysql_version,
+                                        uint64_t max)
 {
-	ut_ad(table->persistent_autoinc);
-	ut_ad(!table->is_temporary());
+  ut_ad(table->persistent_autoinc);
+  ut_ad(!table->is_temporary());
 
-	dict_index_t*	index = dict_table_get_first_index(table);
+  uint64_t autoinc= 0, max_autoinc= 0;
+  dict_index_t *index= dict_table_get_first_index(table);
+  mtr_t mtr;
+  mtr.start();
 
-	if (index == NULL) {
-		return 0;
-	}
+  if (buf_block_t *block=
+      buf_page_get(page_id_t(table->space_id, index->page),
+                   table->space->zip_size(), RW_SX_LATCH, &mtr))
+  {
+    autoinc= page_get_autoinc(block->frame);
 
-	mtr_t		mtr;
-	mtr.start();
-	buf_block_t*	block = buf_page_get(
-		page_id_t(index->table->space_id, index->page),
-		index->table->space->zip_size(),
-		RW_S_LATCH, &mtr);
+    if (autoinc > max || mysql_version == 0 || mysql_version < 100210)
+    {
+      /* Read MAX(autoinc_col), in case this table had originally been
+      created before MariaDB 10.2.4 introduced persistent AUTO_INCREMENT
+      and MariaDB 10.2.10 fixed MDEV-12123, and there could be a garbage
+      value in the PAGE_ROOT_AUTO_INC field. */
+      const dict_col_t *autoinc_col= dict_table_get_nth_col(table, col_no);
+      while (index && index->fields[0].col != autoinc_col)
+        index= dict_table_get_next_index(index);
 
-	ib_uint64_t	autoinc	= block ? page_get_autoinc(block->frame) : 0;
-	const bool	retry	= block && autoinc == 0
-		&& !page_is_empty(block->frame);
-	mtr.commit();
+      if (index)
```

Comment:
Better to do a commit and return if 'index' is not set.
It would make the code more clear instead of relying on that: need_adjust will be false on the next line.
It is likely that you a compiler would in the future give a warning that 'index' may be null at line 1321.

Reply:
It looks like `dict_load_table_one()` would crash if after the call to `dict_load_indexes()` we’d have an empty list `table->indexes`. The condition `if (index)` can therefore be removed.

Reply:
Looks like the index could be NULL in 2 cases.

* Empty index list (not even any cluster index ?)
* Index for the auto increment column is not found.

We don't expect them to happen and possible ways we can handle are ...

1. Check and return Error
2. Assert
3. Remove the checks (should eventually crash if ever happens)
4. Have the checks and skip the evaluation returning zero value (code today)

My preference is same as the list numbering (top to bottom).

Reply:
Sorry, I was mistaken, and we indeed cannot do any further adjustment if `index==nullptr`. This is traversing the list of indexes, trying to find an index that has the `AUTO_INCREMENT` column as the first field. This may indeed fail if the metadata is corrupted in some non-obvious way.